rules/validator.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package rules
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3metrics/filters"
+	"github.com/m3db/m3metrics/metric"
+	"github.com/m3db/m3metrics/policy"
+)
+
+// Validator validates a ruleset.
+type Validator interface {
+	// Validate validates a ruleset.
+	Validate(rs RuleSet) error
+}
+
+type validator struct {
+	opts ValidatorOptions
+}
+
+// NewValidator creates a new validator.
+func NewValidator(opts ValidatorOptions) Validator {
+	return &validator{opts: opts}
+}
+
+func (v *validator) Validate(rs RuleSet) error {
+	// Only the latest (a.k.a. the first) view needs to be validated
+	// because that is the view that may be invalid due to latest update.
+	latest, err := rs.Latest()
+	if err != nil {
+		return NewValidationError(fmt.Sprintf("could not get the latest ruleset snapshot: %v", err))
+	}
+	if err := v.validateMappingRules(latest.MappingRules); err != nil {
+		return NewValidationError(fmt.Sprintf("could not validate mapping rules: %v", err))
+	}
+	if err := v.validateRollupRules(latest.RollupRules); err != nil {
+		return NewValidationError(fmt.Sprintf("could not validate rollup rules: %v", err))
```

Comment:
If the validation issue is a conflict, will this overwrite the conflict error?

Reply:
It will wrap the conflict error.